File: samples/csharp_dotnetcore/48.qnamaker-active-learning-bot/Dialog/DialogHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.AI.QnA;
using Microsoft.Bot.Builder.Dialogs;
using Newtonsoft.Json;

namespace Microsoft.BotBuilderSamples
{
    /// <summary>
    /// QnAMaker Active Learning Dialog helper class
    /// </summary>
    public class DialogHelper
    {
        /// <summary>
        /// QnA Maker Active Learning dialog name
        /// </summary>
        public string ActiveLearningDialogName = "active-learning-dialog";

        /// <summary>
        /// QnA Maker Active Learning Dialog
        /// </summary>
        public WaterfallDialog QnAMakerActiveLearningDialog;

        private QnAMakerOptions qnaMakerOptions;
        private readonly IBotServices _services;

        // Define value names for values tracked inside the dialogs.
        private const string CurrentQuery = "value-current-query";
        private const string QnAData = "value-qnaData";

        // Dialog Options parameters
        private const float DefaultThreshold = 0.03F;
        private const int DefaultTopN = 3;

        // Card parameters
        private const string cardTitle = "Did you mean:";
        private const string cardNoMatchText = "None of the above.";
        private const string cardNoMatchResponse = "Thanks for the feedback.";

        /// <summary>
        /// Dialog helper to generate dialogs
        /// </summary>
        /// <param name="services">Bot Services</param>
        public DialogHelper(IBotServices services)
        {
            QnAMakerActiveLearningDialog = new WaterfallDialog(ActiveLearningDialogName)
                .AddStep(CallGenerateAnswer)
                .AddStep(FilterLowVariationScoreList)
                .AddStep(CallTrain)
                .AddStep(DisplayQnAResult);
            _services = services;
        }

        private async Task<DialogTurnResult> CallGenerateAnswer(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var scoreThreshold = DefaultThreshold;
            var top = DefaultTopN;

            QnAMakerOptions qnaMakerOptions = null;

            // Getting options
            if (stepContext.ActiveDialog.State["options"] != null)
            {
                qnaMakerOptions = stepContext.ActiveDialog.State["options"] as QnAMakerOptions;
                scoreThreshold = qnaMakerOptions?.ScoreThreshold != null ? qnaMakerOptions.ScoreThreshold : DefaultThreshold;
                top = qnaMakerOptions?.Top != null ? qnaMakerOptions.Top : DefaultTopN;
            }

            var response = await _services.QnAMakerService.GetAnswersAsync(stepContext.Context, qnaMakerOptions);

            var filteredResponse = response.Where(answer => answer.Score > scoreThreshold).ToList();

            stepContext.Values[QnAData] = new List<QueryResult>(filteredResponse);
            stepContext.Values[CurrentQuery] = stepContext.Context.Activity.Text;
            return await stepContext.NextAsync(cancellationToken);
        }

        private async Task<DialogTurnResult> FilterLowVariationScoreList(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var response = stepContext.Values[QnAData] as List<QueryResult>;

            var filteredResponse = _services.QnAMakerService.GetLowScoreVariation(response.ToArray()).ToList();

            stepContext.Values[QnAData] = filteredResponse;

            if (filteredResponse.Count > 1)
            {
                var suggestedQuestions = new List<string>();
                foreach (var qna in filteredResponse)
                {
                    suggestedQuestions.Add(qna.Questions[0]);
                }

                // Get hero card activity
                var message = CardHelper.GetHeroCard(suggestedQuestions, cardTitle, cardNoMatchText);

                await stepContext.Context.SendActivityAsync(message);

                return new DialogTurnResult(DialogTurnStatus.Waiting);
            }
            else
            {
                return await stepContext.NextAsync(new List<QueryResult>(response), cancellationToken);
            }
        }

        private async Task<DialogTurnResult> CallTrain(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var trainResponses = stepContext.Values[QnAData] as List<QueryResult>;
            var currentQuery = stepContext.Values[CurrentQuery] as string;

            var reply = stepContext.Context.Activity.Text;

            if (trainResponses.Count > 1)
            {
                var qnaResult = trainResponses.Where(kvp => kvp.Questions[0] == reply).FirstOrDefault();

                if (qnaResult != null)
                {
                    stepContext.Values[QnAData] = new List<QueryResult>() { qnaResult };

                    var records = new FeedbackRecord[]
                    {
                        new FeedbackRecord
                        {
                            UserId = stepContext.Context.Activity.Id,
                            UserQuestion = currentQuery,
                            QnaId = qnaResult.Id,
                        }
                    };

                    var feedbackRecords = new FeedbackRecords { Records = records };

                    // Call Active Learning Train API
                    await _services.QnAMakerService.CallTrainAsync(feedbackRecords);

                    return await stepContext.NextAsync(new List<QueryResult>(){ qnaResult }, cancellationToken);
                }
                else if (reply.Equals(cardNoMatchText))
                {
                    await stepContext.Context.SendActivityAsync(cardNoMatchResponse, cancellationToken: cancellationToken);
                    return await stepContext.EndDialogAsync();
                }
                else
                {
                    return await stepContext.ReplaceDialogAsync(ActiveLearningDialogName, stepContext.ActiveDialog.State["options"], cancellationToken);
                }
            }

            return await stepContext.NextAsync(stepContext.Result, cancellationToken);
        }

        private async Task<DialogTurnResult> DisplayQnAResult(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (stepContext.Result is List<QueryResult> response && response.Count > 0)
            {
                await stepContext.Context.SendActivityAsync(response[0].Answer, cancellationToken: cancellationToken);
            }
            else
            {
                var msg = "No QnAMaker answers found.";
                await stepContext.Context.SendActivityAsync(msg, cancellationToken: cancellationToken);
            }

            return await stepContext.EndDialogAsync();
        }
    }
}

File: samples/javascript_nodejs/48.qnamaker-active-learning-bot/Helpers/dialogHelper.js
/* eslint dot-notation: 0 */
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { CardFactory } = require('botbuilder');
const { DialogTurnStatus, WaterfallDialog } = require('botbuilder-dialogs');

// Card parameters
const cardTitle = 'Did you mean:';
const cardNoMatchText = 'None of the above.';
const cardNoMatchResponse = 'Thanks for the feedback.';

class DialogHelper {
    /**
     * QnAMaker Active Learning Dialog helper class.
     * @param {QnAMaker} qnamaker An instance of QnAMaker service.
     */
    constructor(qnamaker) {
        this.activeLearningDialogName = 'active-learning-dialog';
        this.qnaData = 'value-qnaData';
        this.currentQuery = 'value-current-query';

        this.qnaMaker = qnamaker;
        this.qnaMakerActiveLearningDialog = new WaterfallDialog(this.activeLearningDialogName);

        this.qnaMakerActiveLearningDialog
            .addStep(this.callGenerateAnswer.bind(this))
            .addStep(this.filterLowVariationScoreList.bind(this))
            .addStep(this.callTrain.bind(this))
            .addStep(this.displayQnAResult.bind(this));
    }

    /**
    * @param {WaterfallStepContext} stepContext contextual information for the current step being executed.
    */
    async callGenerateAnswer(stepContext) {
        // Default QnAMakerOptions
        var qnaMakerOptions = {
            ScoreThreshold: 0.03,
            Top: 3
        };

        if (stepContext.activeDialog.state['options'] != null) {
            qnaMakerOptions = stepContext.activeDialog.state['options'];
        }

        // Perform a call to the QnA Maker service to retrieve matching Question and Answer pairs.
        var qnaResults = await this.qnaMaker.getAnswers(stepContext.context, qnaMakerOptions);

        var filteredResponses = qnaResults.filter(r => r.score > qnaMakerOptions.ScoreThreshold);

        stepContext.values[this.qnaData] = filteredResponses;
        stepContext.values[this.currentQuery] = stepContext.context.activity.text;

        return await stepContext.next();
    }

    /**
    * @param {WaterfallStepContext} stepContext contextual information for the current step being executed.
    */
    async filterLowVariationScoreList(stepContext) {
        var responses = stepContext.values[this.qnaData];

        var filteredResponses = this.qnaMaker.getLowScoreVariation(responses);
        stepContext.values[this.qnaData] = filteredResponses;

        if (filteredResponses.length > 1) {
            var suggestedQuestions = [];
            filteredResponses.forEach(element => {
                suggestedQuestions.push(element.questions[0]);
            });

            var message = GetHeroCard(suggestedQuestions, cardTitle, cardNoMatchText);

            await stepContext.context.sendActivity(message);

            return { status: DialogTurnStatus.waiting };
        } else {
            return await stepContext.next(responses);
        }
    }

    /**
    * @param {WaterfallStepContext} stepContext contextual information for the current step being executed.
    */
    async callTrain(stepContext) {
        var trainResponses = stepContext.values[this.qnaData];
        var currentQuery = stepContext.values[this.currentQuery];
        var reply = stepContext.context.activity.text;

        if (trainResponses.length > 1) {
            var qnaResults = trainResponses.filter(r => r.questions[0] === reply);

            if (qnaResults.length > 0) {
                stepContext.values[this.qnaData] = qnaResults;

                var feedbackRecords = {
                    FeedbackRecords: [
                        {
                            UserId: stepContext.context.activity.id,
                            UserQuestion: currentQuery,
                            QnaId: qnaResults[0].id
                        }
                    ]
                };

                // Call Active Learning Train API
                this.qnaMaker.callTrainAsync(feedbackRecords);

                return await stepContext.next(qnaResults);
            } else if (reply === cardNoMatchText) {
                await stepContext.context.sendActivity(cardNoMatchResponse);
                return await stepContext.endDialog();
            } else {
                return await stepContext.replaceDialog(this.activeLearningDialogName, stepContext.activeDialog.state['options']);
            }
        }

        return await stepContext.next(stepContext.result);
    }

    /**
    * @param {WaterfallStepContext} stepContext contextual information for the current step being executed.
    */
    async displayQnAResult(stepContext) {
        var responses = stepContext.result;
        var message = 'No QnAMaker answers found.';

        if (responses != null) {
            if (responses.length > 0) {
                message = responses[0].answer;
            }
        }

        await stepContext.context.sendActivity(message);
        return await stepContext.endDialog();
    }
}

/**
* Get Hero card to get user feedback
* @param {Array} suggestionList A list of suggested questions strings.
* @param {string} cardTitle Title of the card.
* @param {string} cardNoMatchText No match text.
*/
function GetHeroCard(suggestionList, cardTitle = 'Did you mean:', cardNoMatchText = 'None of the above.') {
    var cardActions = [];
    suggestionList.forEach(element => {
        cardActions.push({
            value: element,
            type: 'imBack',
            title: element
        });
    });

    cardActions.push({
        value: cardNoMatchText,
        type: 'imBack',
        title: cardNoMatchText
    });

    var heroCard = CardFactory.heroCard(
        cardTitle,
        [],
        CardFactory.actions(cardActions));

    return { attachments: [heroCard] };
}

module.exports.DialogHelper = DialogHelper;
